oak-run/src/main/java/org/apache/jackrabbit/oak/run/Downloader.java

```diff
@@ -0,0 +1,148 @@
+    private final int connectTimeoutMs;
+    private final int readTimeoutMs;
+    private final List<Future<ItemResponse>> responses;
+
+    public Downloader(int concurrency, int connectTimeoutMs, int readTimeoutMs) {
+        if (concurrency <= 0 || concurrency > 1000) {
+            throw new IllegalArgumentException("concurrency range must be between 1 and 1000");
+        }
+        if (connectTimeoutMs < 0 || readTimeoutMs < 0) {
+            throw new IllegalArgumentException("connect and/or read timeouts can not be negative");
+        }
+        LOG.info("Initializing Downloader with max number of concurrent requests={}", concurrency);
+        this.connectTimeoutMs = connectTimeoutMs;
+        this.readTimeoutMs = readTimeoutMs;
+        this.executorService = new ThreadPoolExecutor(
+                (int) Math.ceil(concurrency * .1), concurrency, 60L, TimeUnit.SECONDS,
+                new LinkedBlockingQueue<>(),
+                new ThreadFactoryBuilder()
+                        .setNameFormat("downloader-%d")
+                        .setDaemon(true)
+                        .build()
+        );
+        this.responses = new ArrayList<>();
+    }
+
+    public void offer(Item item) {
+        Callable<ItemResponse> callableTask = () -> {
+            ItemResponse response = new ItemResponse(item);
+            long t0 = System.nanoTime();
+            try {
+                URLConnection sourceUrl = new URL(item.source).openConnection();
+                sourceUrl.setConnectTimeout(Downloader.this.connectTimeoutMs);
+                sourceUrl.setReadTimeout(Downloader.this.readTimeoutMs);
+
+                Path destinationPath = Paths.get(item.destination);
+                Files.createDirectories(destinationPath.getParent());
+                try (ReadableByteChannel byteChannel = Channels.newChannel(sourceUrl.getInputStream());
+                     FileOutputStream outputStream = new FileOutputStream(destinationPath.toFile())) {
+                    response.size = outputStream.getChannel()
+                            .transferFrom(byteChannel, 0, sourceUrl.getContentLengthLong());
+                }
+            } catch (Exception e) {
+                LOG.error("Error downloading " + item.source + ": " + e.getMessage());
+                response.failed = true;
+                response.throwable = e;
+            } finally {
+                response.time = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - t0);
+            }
+            return response;
+        };
+        responses.add(this.executorService.submit(callableTask));
+    }
+
+    public List<ItemResponse> waitUntilComplete() {
+        return responses.stream()
+                .map(itemResponseFuture -> {
+                    try {
+                        return itemResponseFuture.get();
+                    } catch (InterruptedException | ExecutionException e) {
+                        throw new RuntimeException(e);
+                    }
```

Comment:
It would be nice to have a test that exercises this code path, a download that fails with a RuntimeException, to complement the test where one file fails to download with an IOException. 

More concretely, what happens if during a transfer one of the downloads fails with a RuntimeException? `waitUntilComplete` will abort without calling .get in the remaining tasks. The main program will call shutdown in its finally block, which tries to do an orderly shutdown, waiting for the tasks to complete. But I guess this is not a problem, because shutdown() does not block waiting for the tasks to finish, and the main method will exit after calling shutdown.

Reply:
The only way to exercise this code block is to close the ExecutorService while a download is in progress.

If a single download fails, we will never get an `ExecutionException` because the `callableTask` never throws an Exception. We have two test cases for this:

* `DownloaderTest.downloadMultiWithMissingOne()`
* `DataStoreCopyCommandTest.allBlobsPlusMissingOne`

In both cases, the failed download did not affect the download of the other 2 files.